### PR TITLE
supports https protocol

### PR DESCRIPTION
**Pull request by HTC12587**
supports https protocol

Please review and merge the code

### CODE REVIEW

**Comment by HTC12587** [comment]
@lbajolet-hashicorp 请您审核并合并代码

**Comment by HTC12587** [comment]
@lbajolet-hashicorp Please review and merge the code

**Comment by anshulsharma-hashicorp** [comment]
@HTC12587 there are merge conflict in your PR, coul dyou please resolve that first, will try to check the logic post that thanks.

**Comment by HTC12587** [comment]
> @HTC12587 there are merge conflict in your PR, coul dyou please resolve that first, will try to check the logic post that thanks.

Hello, the conflict has been resolved. Please review and help merge it. Thank you

File: builder/ecs/access_config.go [hunk 1]
```diff
@@ -119,6 +123,11 @@ func (c *AlicloudAccessConfig) Client() (*ClientWrapper, error) {
 
 	client.AppendUserAgent(Packer, version.PluginVersion.FormattedVersion())
 	client.SetReadTimeout(DefaultRequestReadTimeout)
+	if c.Protocol != "" {
```
Comment (anshulsharma-hashicorp) [inline builder/ecs/access_config.go]: these 2 conditions can be combined in one if block with the && logical operator.

File: builder/ecs/access_config_test.go [hunk 2]
```diff
@@ -65,5 +65,10 @@ func TestAlicloudAccessConfigPrepareRegion(t *testing.T) {
 		t.Fatalf("shouldn't have err: %s", err)
 	}
```
Comment (anshulsharma-hashicorp) [inline builder/ecs/access_config_test.go:66]: can we have more unit test cases for negative use cases, where protocol not from the expected set?

File: builder/ecs/access_config.go [hunk 3]
```diff
@@ -138,6 +145,10 @@ func (c *AlicloudAccessConfig) Prepare(ctx *interpolate.Context) []error {
 		errs = append(errs, fmt.Errorf("region option or ALICLOUD_REGION must be provided in template file or environment variables."))
 	}
 
+	if c.Protocol != "" && !isHTTPOrHTTPS(c.Protocol) {
+		errs = append(errs, fmt.Errorf("The template file must provide the correct protocol format, either http or https."))
```
Comment (anshulsharma-hashicorp) [inline builder/ecs/access_config.go:149]: you are not setting the protocol in the config here the way it is happening for http above, like = ```client.GetConfig().WithScheme(strings.ToUpper(c.Protocol))```

Comment (HTC12587) [reply builder/ecs/access_config.go:149]: > you are not setting the protocol in the config here the way it is happening for http above, like = `client.GetConfig().WithScheme(strings.ToUpper(c.Protocol))`

This is an error message reminder returned to the customer, not an assignment

Comment (anshulsharma-hashicorp) [reply builder/ecs/access_config.go:149]: got it, so could you add a test case where the protocol is neither http nor https and assert the expected error message.

**Comment by anshulsharma-hashicorp** [comment]
could you add a test case where the protocol is neither http nor https and assert the expected error message.

**Comment by HTC12587** [comment]
> could you add a test case where the protocol is neither http nor https and assert the expected error message.

I have added a test case with a protocol that is neither HTTP nor HTTPS. Please review it

**Review by anshulsharma-hashicorp — APPROVED** [review]
